### PR TITLE
#166840943 Create role based access control

### PR DESCRIPTION
**Pull request by kechyy**
#### What does this PR do?
- create role based access control functionality

#### Description of Task to be completed?
- create middleware to verify JWT
- create middleware to authorise user types

#### How should this be manually tested?
- clone the repo
- run npm install
- start server at port 3000
- When a user visit a protected route
- The middleware grant access to all authorised user

#### Any background context you want to provide?
- N/A

#### What are the relevant pivotal tracker stories?
- [#166840943](https://www.pivotaltracker.com/story/show/166840943)

#### Screenshots (if appropriate)
 N/A

#### Questions:
- N/A

### CODE REVIEW

File: src/middleware/checkers.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+class Checkers {
```
Comment (timi-codes) [inline src/middleware/checkers.js]: Hello @kechyy, can we have this file renamed?. I think checkers.js is not descriptive enough. Also, appending the folder name to the filename will be nice. Check here for more details. - https://github.com/andela/engineering-playbook/tree/master/5.%20Developing/Conventions#commit-message
I will suggest you use `permission.middleware.js ` or any other name you can think of.

**Review by timi-codes — CHANGES_REQUESTED** [review]

**Review by timi-codes — COMMENTED** [review]
This looks good! @kechyy however I left some comment below.

**Review by Onyimatics — APPROVED** [review]
Nice work @kechyy. I like your initiative on the test file.

File: src/middleware/permission.middleware.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+const userCheck = (req, res, next) => {
+  const {
+    token: { role }
+  } = req;
+  if (role !== 'user' && role !== 'admin' && role !== 'superAdmin') {
```
Comment (segunolalive) [inline src/middleware/permission.middleware.js]:
```
const {
    token: { role }
  } = req;
```
vs
```
const { role } = req.token;
```

File: src/middleware/permission.middleware.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    });
+  }
+  return next();
+};
+
+const adminCheck = (req, res, next) => {
+  const {
+    token: { role }
+  } = req;
+  if (role !== 'admin' && role !== 'superAdmin') {
```
Comment (segunolalive) [inline src/middleware/permission.middleware.js]: switch `negations` to direct comparisons

```
if (role === 'admin' || role === 'superAdmin') {
  return next()
}

return responseGenerator.err(...)
```

File: src/repository/base.repository.js [hunk 4]
```diff
@@ -35,6 +35,33 @@ class BaseRepository {
...
+  /**
+   *
+   *
+   * @static
+   * @param {object} model - database model
+   * @param {object} options - column options
+   * @returns {object} - returns a database object
+   * @memberof BaseRepository
+   */
+  static async delete(model, options) {
```
Comment (segunolalive) [inline src/repository/base.repository.js]: move the `async` modifier to the `controller`

File: src/test/permission.spec.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      role: 'user'
+    };
+    const userToken = jwt.sign(newUser, process.env.JWT_SECRET);
+    chai
+      .request(app)
+      .get('/api/v1/users')
+      .set('x-access-token', userToken)
+      .end((err, res) => {
+        expect(res.status).to.equal(401);
+        expect(res.body).to.have.property('status');
```
Comment (segunolalive) [inline src/test/permission.spec.js]: the response body should not have this property

Comment (segunolalive) [reply src/test/permission.spec.js]: `401` or `403` ?

File: src/test/permission.spec.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        expect(res.body.error).to.equal(
```
Comment (segunolalive) [inline src/test/permission.spec.js]: stick to `message` over `error`

File: src/test/permission.spec.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+          'Unauthorized User, Please contact the administrator.'
+        );
+        done();
+      });
+  });
+
+  it('should return 200 when user is an admin or super admin', done => {
```
Comment (segunolalive) [inline src/test/permission.spec.js]: The test descriptions don't tell what you're testing, making it impossible to tell their value add

**Review by segunolalive — CHANGES_REQUESTED** [review]

File: src/test/user.spec.js [hunk 8]
```diff
@@ -456,3 +457,123 @@ describe('PATCH /api/v1/unfollow', () => {
...
+      .send({ role: 'admin' });
+    expect(res.status).to.equal(403);
+
+    const userRoleStatus = await BaseRepository.findItAll(db.User, {
+      id: secondUser.id
+    });
+    const currentNumberOfUsers = await BaseRepository.findItAll(db.User);
+
+    expect(userRoleStatus[0].role).to.not.equal('admin');
+    expect(userRoleStatus[0].role).to.equal('user');
```
Comment (segunolalive) [inline src/test/user.spec.js]: 😍 😍 😍 😍 😍 

I love the logic of this test.

Please remove the superfluous assertions as discussed.

**Review by segunolalive — CHANGES_REQUESTED** [review]

**Review by timi-codes — APPROVED** [review]
Looks good to me

**Pull request merged**